File: app/src/main/java/com/tcc/sicv/presentation/ProfileViewModel.java
package com.tcc.sicv.presentation;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.tcc.sicv.base.Result;
import com.tcc.sicv.base.ResultListenerFactory;
import com.tcc.sicv.data.firebase.AuthRepository;
import com.tcc.sicv.data.model.FlowState;
import com.tcc.sicv.data.model.User;
import com.tcc.sicv.data.preferences.PreferencesHelper;

import static com.tcc.sicv.data.model.Status.LOADING;

public class ProfileViewModel extends ViewModel {
    private AuthRepository authRepository;
    private MutableLiveData<FlowState<User>> flowState;
    private MutableLiveData<FlowState<Void>> logoutState;
    private PreferencesHelper preferencesHelper;
    private Result<User> mainResultListener;
    private Result<Void> logoutListener;

    public ProfileViewModel(PreferencesHelper preferences) {
        authRepository = new AuthRepository();
        logoutState = new MutableLiveData<>();
        logoutListener = new ResultListenerFactory<Void>().create(logoutState);
        flowState = new MutableLiveData<>();
        mainResultListener = new ResultListenerFactory<User>().create(flowState);
        preferencesHelper = preferences;
        getUserProfile();
    }

    public void logout() {
        logoutState.postValue(new FlowState<Void>(null, null, LOADING));
        preferencesHelper.logout();
        authRepository.logout(logoutListener);
    }

    public void getUserProfile() {
        flowState.postValue(new FlowState<User>(null, null, LOADING));
        authRepository.getUserProfile(preferencesHelper.getEmail(), mainResultListener);
    }

    public LiveData<FlowState<User>> getFlowState() {
        return flowState;
    }

    public LiveData<FlowState<Void>> getLogoutState() {
        return logoutState;
    }
}

File: app/src/main/java/com/tcc/sicv/presentation/ProfileViewModel.kt
package com.tcc.sicv.presentation

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel

import com.tcc.sicv.base.Result
import com.tcc.sicv.base.ResultListenerFactory
import com.tcc.sicv.data.firebase.AuthRepository
import com.tcc.sicv.data.model.FlowState
import com.tcc.sicv.data.model.User
import com.tcc.sicv.data.preferences.PreferencesHelper

import com.tcc.sicv.data.model.Status.LOADING

class ProfileViewModel(private val preferencesHelper: PreferencesHelper) : ViewModel() {
    private val authRepository: AuthRepository = AuthRepository()
    private val flowState: MutableLiveData<FlowState<User>>
    private val logoutState: MutableLiveData<FlowState<Unit>> = MutableLiveData()
    private val mainResultListener: Result<User>
    private val logoutListener: Result<Unit>

    init {
        logoutListener = ResultListenerFactory<Unit>().create(logoutState)
        flowState = MutableLiveData()
        mainResultListener = ResultListenerFactory<User>().create(flowState)
        getUserProfile()
    }

    fun logout() {
        logoutState.postValue(FlowState(null, null, LOADING))
        preferencesHelper.logout()
        authRepository.logout(logoutListener)
    }

    fun getUserProfile() {
        flowState.postValue(FlowState<User>(null, null, LOADING))
        authRepository.getUserProfile(preferencesHelper.email ?:"", mainResultListener)
    }

    fun getFlowState(): LiveData<FlowState<User>> {
        return flowState
    }

    fun getLogoutState(): LiveData<FlowState<Unit>> {
        return logoutState
    }
}
